chore(docs): Rendering Options Conceptual Guide (#33129)
gatsbyjs/gatsby: docs/docs/conceptual/rendering-options.md

---
title: Rendering Options
---

Gatsby is historically known as a static site generator enhanced with [React Hydration][1].
But starting with Gatsby 4, you can choose alternative rendering modes in addition to static site generation (SSG).
Now Gatsby covers all possible use-cases you may face as a web developer.

### What is a rendering mode?

Rendering mode defines at what stage the user-facing HTML is generated. It can happen at build time
(SSG or pre-rendering), during HTTP request (server-side rendering) or locally in the browser
with Javascript (client-side rendering).

> For an in-depth explanation of each of those approaches and their trade-offs we
> highly recommend the ["Rendering on the Web"][3] article from the Chrome team.

As already mentioned, Gatsby always supported SSG and [client-side rendering][4], and now the two
other rendering options are available: Deferred Static Generation (DSG) and Server-side rendering (SSR).
Let's dive into what each of those modes mean.

## Static Site Generation (SSG)

> SSG is the default rendering mode in Gatsby. Gatsby generates all assets and HTML for all pages at build time on a

Reviewer: this probably need better wording than my suggestion, but "for all pages" is the "old" way, and now it "doesn't have to be all pages at build time"

```suggestion
SSG is the default rendering mode in Gatsby. Gatsby generates all assets and HTML for all SSG pages at build time on a
```

build server (this could be your laptop, any build service, or Gatsby Cloud worker if you use [Gatsby Cloud][6]).

> Then, the produced static files are uploaded to the CDN provider of choice and served to end-users.

Reviewer 2: ```suggestion
Then, the produced static files are uploaded to the content delivery network (CDN) provider of choice and served to end users.
```

The build server is not needed after the build step and could even be turned off.

This mode provides the most pleasant user experience, the highest level of security, and run-time scalability for your site.

> **Note:** SSG doesn't mean your site is not dynamic. You can still use JavaScript to communicate with any APIs,
> add private sections of your site for authorized users via [client-side rendering][4] and

> > have any features SPA can have

Reviewer 2: ```suggestion
> have any features that single-page applications (SPAs) can have.
```

> Here is how SSG works in Gatsby Cloud (the principle remains the same with any build and CDN provider):

Reviewer 2: ```suggestion
Here is how SSG works in Gatsby Cloud (although the principle remains the same with any build and CDN provider):
```

![Static Site Generation diagram](../images/ssg-diagram.jpg)

> One downside of SSG is building time. As the number of pages of your site grows, so does the build time.
> Gatsby supports [incremental builds][5] to make sure the 2nd and subsequent builds only rebuild what was changed
> but for the initial build (the build without the cache), build times may become an issue. That's where the
> Deferred Static Generation could be beneficial.

Reviewer 2: ```suggestion
One downside of SSG is longer build times. As the number of pages of your site grows, so does the build time.
Gatsby supports [incremental builds][5] to make sure the 2nd and subsequent builds only rebuild the parts of your site that changed,
but for the initial build (the build without the cache), build times may become an issue. That's where
Deferred Static Generation could be beneficial.
```

## Deferred Static Generation (DSG)

As the title suggests, Deferred Static Generation is conceptually very similar to SSG. The only difference is that some
pages are not generated at build time and instead generated during the first HTTP request to this page.

The subsequent requests will hit the CDN cache the same way as with SSG.

For example: maybe you have an archive of old articles that no longer receive significant traffic. There is

> no practical reason to generate them on each build and thus delay the delivery of fresh articles.

Reviewer 2: ```suggestion
no practical reason to generate them on each build (and thus delay the delivery of fresh articles).
```

In this case, you may choose to defer the generation of old pages, and Gatsby will skip them during the build step.

> - [How to use deferred static generation][7]

Reviewer 2: ```suggestion
- [How-To Guide: Using Deferred Static Generation][7]
```

![Deferred Static Generation diagram](../images/dsg-diagram.jpg)

Unlike SSG, deferred static generation requires you to keep the build server running after the initial build.
It implies a different deployment model and requires backend infrastructure.

But don't worry, [Gatsby Cloud][6] supports it out-of-the-box.

> ## Server-side Rendering (SSR)

Reviewer 2: ```suggestion
## Server-Side Rendering (SSR)
```

> SSG, DSG, and client-side rendering can handle a vast majority of use-cases in web development.

Reviewer 2: ```suggestion
SSG, DSG, and client-side rendering can handle a vast majority of use cases in web development.
```

But there is a small niche when you may still need to generate HTML on-the-fly.

For example: maybe you are building a site with user reviews and want those reviews to be

> immediately indexed by search engines as soon as they are posted. So the client-side rendering is not an option.

Reviewer: bit rusty on this, but don't search engines (particularly google) will actually use js, so indexing still would happen?

Reviewer 3: It is being indexed but JS based sites are slower

Reviewer 3: Perhaps we should word it to something along these lines:

```
For example: you have a personalised landing page for each user. Getting the data at server time can reduce the page from shifting its layout while data gets loaded. Therefore, it can give a better user experience than client side only rendering.

Reviewer 4: Yeah, according to [google docs](https://developers.google.com/search/docs/advanced/javascript/javascript-seo-basics) google bot does run JS. But not sure if it is the same for all search engines (Bing, Yandex, etc).

But even those docs contain the following note:

> Keep in mind that server-side or pre-rendering is still a great idea because it makes your website faster for users and crawlers, and not all bots can run JavaScript.

In this case, you may choose server-side rendering for pages with user reviews. Any time someone

> requests this page, Gatsby will call `getServerData` function defined in the page component.

Reviewer 2: ```suggestion
requests this page, Gatsby will call the `getServerData` function defined in the page component.
```

> That's where you can request any data from 3rd-party APIs. Gatsby passes the returned result as `serverData`

Reviewer 2: ```suggestion
That's where you can request any data from 3rd-party APIs. Gatsby passes the returned result as a `serverData`
```
